lib/metasploit/framework/login_scanner/snmp.rb

```diff
@@ -54,93 +60,286 @@ def versions
+            resend_count += 1
+            if resend_count > MAX_RESEND_COUNT
+              return false
+            end
+            ::IO.select(nil, nil, nil, 0.25)
+            retry
+          rescue ::Rex::ConnectionError
+            # This fires for host unreachable, net unreachable, and broadcast sends
+            # We can safely ignore all of these for UDP sends
           end
-          proof
         end
 
+        # Create a SNMP request that tries to read from sys.sysDescr.0
+        def create_snmp_read_sys_descr_request(version_str, community)
+          version = version_str == '1' ? 1 : 2
+          OpenSSL::ASN1::Sequence([
+            OpenSSL::ASN1::Integer(version - 1),
+            OpenSSL::ASN1::OctetString(community),
+            OpenSSL::ASN1::Set.new([
+              OpenSSL::ASN1::Integer(rand(0x80000000)),
+              OpenSSL::ASN1::Integer(0),
+              OpenSSL::ASN1::Integer(0),
+              OpenSSL::ASN1::Sequence([
+                OpenSSL::ASN1::Sequence([
+                  OpenSSL::ASN1.ObjectId("1.3.6.1.2.1.1.1.0"),
+                  OpenSSL::ASN1.Null(nil)
+                ])
+              ]),
+            ], 0, :IMPLICIT)
+          ]).to_der
+        end
+
+        # Create a SNMP request that tries to write to sys.sysDescr.0
+        def create_snmp_write_sys_descr_request(version_str, community, data)
+          version = version_str == '1' ? 1 : 2
+          snmp_write = OpenSSL::ASN1::Sequence([
+            OpenSSL::ASN1::Integer(version - 1),
+            OpenSSL::ASN1::OctetString(community),
+            OpenSSL::ASN1::Set.new([
+              OpenSSL::ASN1::Integer(rand(0x80000000)),
+              OpenSSL::ASN1::Integer(0),
+              OpenSSL::ASN1::Integer(0),
+              OpenSSL::ASN1::Sequence([
+                OpenSSL::ASN1::Sequence([
+                  OpenSSL::ASN1.ObjectId("1.3.6.1.2.1.1.1.0"),
+                  OpenSSL::ASN1::OctetString(data)
+                ])
+              ]),
+            ], 3, :IMPLICIT)
+          ]).to_der
+        end
+
+        # Parse a SNMP reply from a packet and return a response hash or nil
+        def parse_snmp_response(pkt)
+          asn = OpenSSL::ASN1.decode(pkt) rescue nil
+          return if not asn
+
+          snmp_vers  = asn.value[0].value.to_i rescue nil
+          snmp_comm  = asn.value[1].value rescue nil
+          snmp_error = asn.value[2].value[1].value.to_i rescue nil
+          snmp_data  = asn.value[2].value[3].value[0] rescue nil
+          snmp_oid   = snmp_data.value[0].value rescue nil
+          snmp_info  = snmp_data.value[1].value.to_s rescue nil
+
+          return if not (snmp_error and snmp_comm and snmp_data and snmp_oid and snmp_info)
+          snmp_vers = snmp_vers == 0 ? "1" : "2c"
+
+          { error: snmp_error, community: snmp_comm, proof: snmp_info, version: snmp_vers}
+        end
 
+        # Create a new socket for this scanner
+        def configure_socket
```

Comment:
TCP is not that common for SNMP. We might want to add the option in the future, but it would require a full refactor of the current implementation to do TCP instead of UDP. Also, slower, especially the service disconnects after each bad community